### DISCUSSION TITLE
Question about streaming plugin and new rack developments

### DISCUSSION THREAD

**Discussion by cpgo** [jeremyevans/roda#251] [General]
Hello, forgive me if this is a dumb question as I am not very familiar with the subject.

How does the discussion here https://github.com/rack/rack/pull/1745 affects roda and the streaming plugin? Did roda already had its own implementation independent of rack to make SSE work?

And another question, do I need any special setup on the puma side to make SSE work? I could find some example snippets by jeremy on the google group but when I try to implement it the response seem to be buffered and not streamed.

On [this repo](https://github.com/cpgo/roda-multi-turbo)  I'm playing around with various roda plugins an trying to integrate it with hotwire/turbo.
The streaming part is more or less working, i can see the messages streamed _but_ i cant inspect the event stream on firefox dev tools, and I cant contron-c the puma server, if I wanna restart the app I have to kill the puma process.

Thanks everyone for this awesome project!

**Comment by jeremyevans** [comment]
That rack PR makes no difference to Roda.  The Roda streaming plugin uses Rack's support for enumerable bodies for streaming, it has no need to use the new support for callable bodies (which is just the same response hijacking rack has supported for many years, in a slightly nicer API).

I don't think you should need anything on the puma side to get SSE to work, but it's hard to comment unless you provide a small example of what you are doing.

**Reply by cpgo** [reply]
Got it, thanks Jeremy.
As for what I am doing, this is the streaming route I have on the linked repo
```ruby
# app entrypoint
# frozen_string_literal: true

require "roda"

class App < Roda
  plugin :render, layout: "./layout"
  plugin :assets
  plugin :view_options
  plugin :hash_routes
  plugin :header_matchers
  plugin :json
  plugin :streaming

  Dir["routes/*.rb"].each { |f| require_relative f }

  route do |r|
    r.hash_routes("")
  end
end
```

```ruby
#config.ru
require_relative "app"

run App.freeze.app
```

```ruby
# route file with streaming route
# frozen_string_literal: true

class App
  hash_path "/messages" do |r|
    r.on accept: "text/html" do
      sleep 5 # pretend its a long running query
      view("messages")
    end
    r.on accept: "text/vnd.turbo-stream.html" do
      render("messages")
    end
  end

  hash_path "/messages/stream" do |r|
    r.get do
      response["Content-Type"] = "text/event-stream; utf-8"
      response["Cache-Control"] = "no-cache"
      response["Transfer-Encoding"] = "chuncked"
      response["X-Accel-Buffering"] = "no"

      stream(loop: true) do |out|
        %w[a b c].each do |v|
          message = "event: custom-event\n" + "id: #{rand}\n" + "data: #{v}\n\n"
          out << message
          sleep 1
        end
      end
    end
  end
end

```
And on the client part
```javascript
      <script>
        (function() { "use strict";
            var ul = document.querySelector('ul');
            var es = new EventSource('/messages/stream');
            function li(text) {
                var li = document.createElement('li');
                li.innerText = text;
                ul.appendChild(li);
            }
            es.addEventListener('open', function() {
                li('Server connected :)');
            });
            es.addEventListener('custom-event', function(event) {
                li(event.data);
            });
            es.addEventListener('error', function() {
                li('Server unavailable :(');
            });
        })();
    </script>
```
I see two strange behaviours:

1.  It looks like firefox dev tools dont handle SSE very well, I can see the initiated request but I cant inspect the events comming in

    1.1  On chrome/edge I can inspect the events normally
    1.2 The js client works on both firefox and chrome, maybe the network dev tools is confused because im returning wrong headers and/or format

2. If I try to exit the server (started with `bundle exec rackup`) by pressing control-c the server simply doest quit, I have to `kill -9` the puma process manually.

I am using puma default configs as I dont have a config file for that right now.

**Reply by jeremyevans** [reply]
The one bug I noticed is that `chunked` is misspelled in the `Transfer-Encoding` header.  You can probably drop the `; utf-8` from the `Content-Type`, since event streams should already been required to be in UTF-8 (https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events#event_stream_format). 

Looks like your issue with 1. is specific to Firefox.  Maybe that is because of the header issues, I'm not sure.  For 2., if you press Control-C, then close the browser or navigate to a different page, does puma stop?  Puma's documentation on signals (https://github.com/puma/puma/blob/master/docs/signals.md) says that it will `attempt to finish then exit`, which my guess means it will not accept new connections, but will not exit until it has finished processing existing connections.

**Reply by cpgo** [reply]
About `2.`,  if I close the broser tab before exiting the server it works as expected.

There is definetly some weird stuff going in with my headers
If I let a typo on `Transfer-Encoding` the streaming works

Chromium is on the left, FF is on the right

https://user-images.githubusercontent.com/739891/155374413-bd10db0f-9b9a-4530-88e1-e797b1238b8a.mp4


If I fix the header value the stream dont work on chromium


https://user-images.githubusercontent.com/739891/155374430-9653f7c0-8759-4dd6-b265-7ef0c74442fb.mp4

I will investigate it a little better later, thanks for the replies. If you want o look at the source code I'm using the repo is linked on my first message.
